common/xdc/nDCHistoryResender.go

```diff
@@ -0,0 +1,217 @@
+	NDCHistoryRereplicatorImpl struct {
+		targetClusterName    string
+		domainCache          cache.DomainCache
+		adminClient          a.Client
+		historyReplicationFn nDCHistoryReplicationFn
+		serializer           persistence.PayloadSerializer
+		replicationTimeout   time.Duration
+		logger               log.Logger
+	}
+)
+
+// NewNDCHistoryRereplicator create a new NDCHistoryRereplicatorImpl
+func NewNDCHistoryRereplicator(
+	targetClusterName string,
+	domainCache cache.DomainCache,
+	adminClient a.Client,
+	historyReplicationFn nDCHistoryReplicationFn,
+	serializer persistence.PayloadSerializer,
+	replicationTimeout time.Duration,
+	logger log.Logger,
+) *NDCHistoryRereplicatorImpl {
+
+	return &NDCHistoryRereplicatorImpl{
+		targetClusterName:    targetClusterName,
+		domainCache:          domainCache,
+		adminClient:          adminClient,
+		historyReplicationFn: historyReplicationFn,
+		serializer:           serializer,
+		replicationTimeout:   replicationTimeout,
+		logger:               logger,
+	}
+}
+
+// SendSingleWorkflowHistory sends one run IDs's history events to remote
+func (n *NDCHistoryRereplicatorImpl) SendSingleWorkflowHistory(
+	domainID string,
+	workflowID string,
+	runID string,
+	startEventID *int64,
+	startEventVersion *int64,
+	endEventID *int64,
+	endEventVersion *int64,
+) error {
+
+	var token []byte
+	for doPaging := true; doPaging; doPaging = len(token) > 0 {
+		response, err := n.getHistory(
+			domainID,
+			workflowID,
+			runID,
+			startEventID,
+			startEventVersion,
+			endEventID,
+			endEventVersion,
+			token,
+			defaultPageSize)
+		if err != nil {
+			return err
+		}
+		token = response.NextPageToken
+
+		// we don't need to handle continue as new here
+		// because we only care about the current run
+		// TODO: revisit to evaluate if we need to handle continue as new
+		for _, batch := range response.HistoryBatches {
+			replicationRequest := n.createReplicationRawRequest(
+				domainID,
+				workflowID,
+				runID,
+				batch,
+				response.GetVersionHistory().GetItems())
+			err := n.sendReplicationRawRequest(replicationRequest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (n *NDCHistoryRereplicatorImpl) createReplicationRawRequest(
+	domainID string,
+	workflowID string,
+	runID string,
+	historyBlob *shared.DataBlob,
+	versionHistoryItems []*shared.VersionHistoryItem,
+) *history.ReplicateEventsV2Request {
+
+	request := &history.ReplicateEventsV2Request{
+		DomainUUID: common.StringPtr(domainID),
+		WorkflowExecution: &shared.WorkflowExecution{
+			WorkflowId: common.StringPtr(workflowID),
+			RunId:      common.StringPtr(runID),
+		},
+		Events:              historyBlob,
+		VersionHistoryItems: versionHistoryItems,
+	}
+
+	return request
+}
+
+func (n *NDCHistoryRereplicatorImpl) sendReplicationRawRequest(
+	request *history.ReplicateEventsV2Request,
+) error {
+
+	if request == nil {
+		return nil
+	}
```

Comment:
when will the request be nil?

Reply:
Some special logic in 2DC rereplicator. remove the check